File: Week5/2 - String Valid in Binary Tree/Solution.py
#
# Created on Thu Apr 30 2020
#
# Title: Leetcode - Check If a String Is a Valid Sequence from Root to Leaves Path in a Binary Tree
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def is_valid(self, root, arr, index):
        if root.val != arr[index]:
            return False
        if index == len(arr) - 1:
            return not(root.left) and not(root.right)
        return ((root.left and self.is_valid(root.left, arr, index+1)) or (root.right and self.is_valid(root.right, arr, index+1)))
               

    def isValidSequence(self, root, arr):
        if root is None or len(arr) == 0:
            return False
        return self.is_valid(root, arr, 0)

File: Week5/2 - String Valid in Binary Tree/Solution.cpp
/*
 * Created on Thu Apr 30 2020
 *
 * Title: Leetcode - Check If a String Is a Valid Sequence from Root to Leaves Path in a Binary Tree
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node. 
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    bool valid_path_using_dfs(TreeNode* root, vector<int>& arr, int idx) {
        if(root->val != arr[idx]) return false;

        if(idx == arr.size()-1) return !root->left && !root->right;
        
        return ((root->left && valid_path_using_dfs(root->left, arr, idx+1)) || (root->right && valid_path_using_dfs(root->right, arr, idx+1))); 
    }
    bool isValidSequence(TreeNode* root, vector<int>& arr) {
        if(!root) return arr.size() == 0;
        
        return valid_path_using_dfs(root, arr, 0);
    }
};